File: 1018 Public Bike Management.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <tuple>
using namespace std;
int capacity, n, endIdx, roadSize;
vector<vector<int>> gragh, pre;
deque<int> stationWeights, path, temppath, dist;
int minNeed = INT16_MAX, minBack = INT16_MAX;
void DFS(int v)
{
    temppath.push_front(v);
    for (auto value : pre[v])dfs(value);
    if (pre[v].empty())
    {
        int need = 0, back = 0;
        for (auto && idx : temppath)
        {
            back += stationWeights[idx];
            if (back<0)
            {
                need += -back;
                back = 0;
            }
        }
        if (need < minNeed || (need == minNeed && minBack>back))
            tie(path, minNeed, minBack) = { temppath,need,back };
    }
    temppath.pop_front();
}
int main()
{
    cin >> capacity >> n >> endIdx >> roadSize;
    gragh.assign(n + 1, vector<int>(n + 1, INT16_MAX));
    dist.resize(gragh.size(), INT16_MAX);
    stationWeights.resize(gragh.size());
    for (int i = 1; i < gragh.size(); ++i)
    {
        cin >> stationWeights[i];
        stationWeights[i] -= capacity / 2;
    }
    for (int i = 0; i < roadSize; ++i)
    {
        int start, end, time;
        cin >> start >> end >> time;
        gragh[start][end] = time;
        gragh[end][start] = time;
    }
    pre.assign(gragh.size(), {});
    deque<bool> collected(gragh.size());
    auto dist = gragh[0];
    dist[0] = 0;
    while (true)
    {
        int idx = -1;
        int min = INT16_MAX;
        for (int i = 0; i < gragh.size(); ++i)
            if (!collected[i] && dist[i] < min)tie(idx, min) = { i,dist[i] };
        if (idx == -1)break;
        collected[idx] = true;
        for (int i = 0; i < gragh.size(); ++i)
        {
            auto weight = gragh[idx][i];
            if (dist[i]>dist[idx] + weight)
            {
                dist[i] = dist[idx] + weight;
                pre[i].assign({ idx });
            }
            else if (dist[i] == dist[idx] + weight)pre[i].push_back(idx);
        }
    }
    DFS(endIdx);
    printf("%d 0", minNeed);
    for (int i = 1; i < path.size(); i++)printf("->%d", path[i]);
    printf(" %d", minBack);
}

File: 1018 Public Bike Management .py
from itertools import *

max_capacity, station_size, problem_idx, road_size = map(int, input().split())
stations_bike_size = list(chain([0], map(int, input().split())))
graph = [[1e9 for i in range(station_size + 1)] for i in range(station_size + 1)]
dist = [1e9 for i in range(station_size + 1)]
dist[0] = 0
for i in range(road_size):
    start, end, time = map(int, input().split())
    graph[start][end] = time
    graph[end][start] = time
not_visited = set(i for i in range(station_size + 1))
pre_nodes = [[] for i in range(station_size + 1)]
while not_visited:
    min_idx = min(not_visited, key=lambda i: dist[i])
    not_visited.remove(min_idx)
    for idx in not_visited:
        if dist[idx] > dist[min_idx] + graph[min_idx][idx]:
            dist[idx] = dist[min_idx] + graph[min_idx][idx]
            pre_nodes[idx] = [min_idx]
        elif dist[idx] == dist[min_idx] + graph[min_idx][idx]:
            pre_nodes[idx].append(min_idx)
path = []
min_need = 1e9
min_surplus = 1e9
perfect_path = None


def dfs(idx):
    global path, min_need, min_surplus, perfect_path
    if idx == 0:
        need, surplus = 0, 0
        for idx in reversed(path):
            surplus += stations_bike_size[idx] - max_capacity // 2
            if surplus < 0:
                need -= surplus
                surplus = 0
        if need < min_need or (need == min_need and surplus < min_surplus):
            min_need, min_surplus, perfect_path = need, surplus, path[::-1]
    else:
        path.append(idx)
        for i in pre_nodes[idx]:
            dfs(i)
        path.pop()


dfs(problem_idx)
print(min_need, '->'.join(chain('0', (str(i) for i in perfect_path))), min_surplus)
